### PR TITLE
Introduce Spicy parser framework with HTTP parser implementation

### PR DESCRIPTION
**Pull request by Boolean-Autocrat**
Implements 2/3 of #184 

This PR introduces Spicy protocol parser integration with C++ into Glutton, along with a concrete Spicy HTTP parser and handler which mimics the existing functionality of the pure-Go HTTP parser.

Spicy parsers can be enabled via Viper config (`spicy.enabled: true`) and will seamlessly replace the Go parsers for the supported protocols (only HTTP for now).

### C++ Bridge Layer (`protocols/spicy/bridge.{h,cpp}`)
The core integration uses a C++ bridge that interfaces between Spicy/HILTI runtimes and Go:

- Runtime Management: Thread-safe initialization and cleanup of Spicy/HILTI runtimes
- Generic Parsing Interface: `spicy_parse_generic()` accepts any parser name and raw data
- Memory Management: Safe allocation/deallocation with a lot of OOM protection (added in later commits)
- Data Structure Flattening: Recursive traversal of HILTI type system to flatten nested structures into flat key-value pairs using dot notation (e.g., version.number, headers[0].name)

### Go Worker Thread Model (`protocols/spicy/parser.go`)
This was a fairly important design decision I had to make. Due to Spicy/HILTI's thread-local storage requirements, I decided that parsing should operate through a dedicated worker:
- Single Worker Thread: Locked to OS thread with `runtime.LockOSThread()`
- Command Channel: Thread-safe communication via `workerCmd` messages
- Parser Registry: Automatic discovery and registration of available Spicy parsers
- Timeout Protection, among other things

### Key C++ Parsing Function
The `dump_value()` function recursively processes HILTI's type system:
- Containers: Arrays/vectors become `field[0]`, `field[1]`, etc.
- Maps: Key-value pairs become `prefix.key`
- Structs: Fields become `prefix.fieldname`
- Depth Limiting: Prevents stack overflow with 64-level limit
- Binary Data: Automatically detects and preserves binary vs. text data

## Implementation
The new HTTP grammar file (`protocols/spicy/parsers/http.spicy`) shows the integration of Spicy into the Glutton system. I've also written a corresponding handler (`protocols/spicy/handlers/http.go`) which tries to maintain behavioral compatibility with the existing pure-Go HTTP handler (while leveraging Spicy under the hood).

As for the integration points, in `protocols/protocols.go` I have added a simple `viper.GetBool("spicy.enabled")` within the "poor man's detection" for HTTP and the entire Spicy runtime management (init and shutdown) happens in `glutton.go` after checking the same Viper boolean.

## Adding New Parsers
I have tried to keep the addition of new parsers as seamless as possible. The flow looks something like this:
- Write Spicy Grammar: Create `.spicy` file in `protocols/spicy/parsers/`
- Build Parser: `make spicy` automatically compiles Spicy grammars to C++
- Create Handler: Implement Go handler in `protocols/spicy/handlers/`
- Auto-Discovery: Parser is automatically registered and available

Other than the usual change for registering in `protocols/protocols.go` everything can basically remain untouched. The only change required in `protocols/spicy/bridge.cpp` will be the addition of an include statement for the new parser's generated header file.

## Notes
- For now, I am not shipping any of the `.cc` or `.h` files generated by Spicy. In the main Glutton directory, you can simply run `make spicy` which will trigger their creation.
- As for testing, I have added two simple test files `protocols/spicy/parser_test.go` and `protocols/spicy/handlers/http_test.go` which verify the workings of both the parser and the handler with some basic edge cases.
- Build times for the Go binary are a bit longer now due to the inclusion of Spicy and its generated C++ files
- A lot of the intuition for this implementation was derived from the Silk PoC, so thanks for that :)
- I've tried to leave relevant comments in all critical places explaining the "why" of the way specific things were done
- I've also tried my best to cover all C++ edge cases and implement measures at points of failure, but I may have missed something so I'm open to making any changes to the implementation.

## Screenshots (showing the Spicy HTTP parser in action)

Commands ran:
```
curl http://localhost:5000/test

# triggers handleEthereumRPC
curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","method":"eth_blockNumber","params":[],"id":1}' http://localhost:5000/rpc

# triggers handleWallet
curl http://localhost:5000/wallet/balance

# triggers handleDockerAPIVersion
curl http://localhost:5000/v1.16/version

# triggers handleCitrixSMB
curl http://localhost:5000/vpn/index.html

# triggers handleVMwareSend
curl -X POST -H "Content-Type: text/plain" -d "test data with 127.0.0.1 22 embedded" http://localhost:5000/sdk/hyper/send
```

Glutton Logs 1
Glutton Logs 2

### CODE REVIEW

**Comment by Boolean-Autocrat** [comment]
@glaslos @furusiyya Here is my PR for the protocol parser implementation.

I believe the GitHub action will need to be changed to accommodate the Spicy runtime. I've left `spicy.enabled` to `true` by default. The `cgo` flags in `protocols/spicy/parser.go` are Linux specific, so I hope that isn't an issue.

We might need to include some documentation about the installation of the Spicy runtime and clang too.

**Comment by Boolean-Autocrat** [comment]
@glaslos @furusiyya I have sorted out the GitHub actions checks too. The checks are passing successfully after including the Spicy runtime in the actions runner.

File: Makefile [hunk 1]
```diff
@@ -20,7 +20,10 @@ upx:
 default: build
 
 build:
-	go build -ldflags=$(LDFLAGS) -o bin/server app/server.go
+	CC=clang CXX=clang++ go build -o bin/server app/server.go
```
[the remark this answers is not in the document]
Comment (Boolean-Autocrat) [reply Makefile]: This was a minor oversight at my end. I accidentally skipped the linker's existing flags.

[the remark this answers is not in the document]
Comment (Boolean-Autocrat) [reply Makefile]: I think it's safe to bring in copilot's commit here.

File: protocols/spicy/handlers/http.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	return true
+}
+
+func handleDockerAPIVersion(uri string, conn net.Conn) bool {
+	if !strings.HasPrefix(uri, "/v1.16/version") {
+		return false
+	}
+	if data, err := tcp.Res.ReadFile("resources/docker_api.json"); err == nil {
+		conn.Write(append([]byte(fmt.Sprintf(
+			"HTTP/1.1 200 OK\r\nContent-Length:%d\r\n\r\n", len(data))), data...))
```
[the remark this answers is not in the document]
Comment (Boolean-Autocrat) [reply protocols/spicy/handlers/http.go]: Latest commit (51ed6f2) solves this.

File: protocols/spicy/parser.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+	cmdCh <- workerCmd{kind: cmdParse, parser: name, data: data, replyChan: resp}
+
+	var raw any
+	select {
+	case raw = <-resp: // normal path
+	case <-time.After(parseTimeout): // worker stalled
+		// drain resp to free C memory even after we have returned
+		go func() {
+			if raw := <-resp; raw != nil {
+				if p, ok := raw.(*C.ParsedData); ok {
```
[the remark this answers is not in the document]
Comment (Boolean-Autocrat) [reply protocols/spicy/parser.go]: This seems fair. Adding copilot's minor changes to the PR.

**Comment by Boolean-Autocrat** [comment]
@glaslos I've resolved all of Copilot's concerns; they were fairly minor.

I hope the overall structure of the implementation is fine. It would be great if you could go through `protocols/spicy/bridge.cpp` once; it's the main bridge layer and Copilot didn't comment on it either.

File: .gitignore [hunk 4]
```diff
@@ -37,3 +37,7 @@ poc/
 
 # Dev
 .vscode
+
+# Spicy generated files
+protocols/spicy/*.cc
+protocols/spicy/parsers/*.h
```
Comment (glaslos) [inline .gitignore:43]: I think this makes a lot of sense

File: protocols/spicy/parser.go
Comment (glaslos) [inline protocols/spicy/parser.go:1]: Did you consider introducing a `type Parser` struct that holds the state instead of using package global values?

File: protocols/spicy/parser.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+// analyzes protocol data using the appropriate Spicy parser
+// the parser is automatically selected based on the protocol name
+func Parse(proto string, data []byte) (*ParsedData, error) {
+	parsersMutex.RLock()
+	name, ok := registeredParsers[strings.ToLower(proto)] // parser lookup
+	parsersMutex.RUnlock()
+	if !ok {
+		return nil, fmt.Errorf("no Spicy parser registered for %q", proto)
+	}
+	name = strings.TrimSpace(name)
```
Comment (glaslos) [inline protocols/spicy/parser.go]: should be done before looking up the parser.

**Review by glaslos — CHANGES_REQUESTED** [review]
Some initial thoughts.

File: .gitignore (hunk 4 above)
Comment (Boolean-Autocrat) [reply .gitignore:43]: I honestly wanted both sets of generated files to be in the same folder, but it was leading to a ton of compilation issues. I suppose this works as well then 👍

File: protocols/spicy/parser.go
Comment (Boolean-Autocrat) [reply protocols/spicy/parser.go:1]: I'll look into this and fix the implementation.

File: protocols/spicy/parser.go (hunk 5 above)
Comment (Boolean-Autocrat) [reply protocols/spicy/parser.go]: Fixed in latest commit (4001b4a).

File: protocols/spicy/parser.go
Comment (glaslos) [reply protocols/spicy/parser.go:1]: Did you had a chance to look into this?

**Comment by furusiyya** [comment]
@glaslos just for your info:

Mid-term Review Meeting Minutes:
Suyash will integrate Lukas' comments and work on the Spicy DNS parser and its test cases next.

July Week 5 Sync:
Will be wrapping it up this week.

**Comment by glaslos** [comment]
@furusiyya what is needed to get this over the finish line? In case @Boolean-Autocrat is not available anymore, I can get the PR ready to be merged.

**Comment by furusiyya** [comment]
@glaslos I was waiting for @Boolean-Autocrat to push spicy DNS parser but he is not available. I will try to get this over the finish line in next month.
